x/merkledb/history.go

```diff
@@ -160,26 +161,32 @@ func (th *trieHistory) getValueChanges(startRoot, endRoot ids.ID, start, end []b
 					(len(endPath) == 0 || key.Compare(endPath) <= 0) {
 					if existing, ok := combinedChanges.values[key]; ok {
 						existing.after = valueChange.after
+						if existing.before.hasValue == existing.after.hasValue &&
+							bytes.Equal(existing.before.value, existing.after.value) {
+							delete(combinedChanges.values, key)
+							sortedKeys.Delete(key)
+						}
 					} else {
 						combinedChanges.values[key] = &change[Maybe[[]byte]]{
 							before: valueChange.before,
 							after:  valueChange.after,
 						}
+						sortedKeys.ReplaceOrInsert(key)
 					}
-					sortedKeys.ReplaceOrInsert(key)
-				}
-			}
-
-			// Keep only the smallest [maxLength] items in [combinedChanges.values].
-			for sortedKeys.Len() > maxLength {
-				if greatestKey, found := sortedKeys.DeleteMax(); found {
-					delete(combinedChanges.values, greatestKey)
 				}
 			}
 
 			return true
 		},
 	)
+
+	// Keep only the smallest [maxLength] items in [combinedChanges.values].
+	for sortedKeys.Len() > maxLength {
```

Comment:
move this outside of the loop since a no-op key/value could cause a value that was preiously excluded to be included again.